From 6717121b5eec8fc5beac9c783a3db3e0422d69c7 Mon Sep 17 00:00:00 2001
From: Sujeet Baranwal <sbaranwal@marvell.com>
Date: Fri, 24 Jul 2020 15:20:40 -0700
Subject: [PATCH 1214/1239] smbios: type 9 implementation

All system slots are described in SMBIOS type 9 as added
in this patch.

Signed-off-by: Sujeet Baranwal <sbaranwal@marvell.com>
Change-Id: Iac31be47931df50e486c6c052cf8adbed3b44f3f
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/boot/u-boot/+/35166
Reviewed-by: Jia Ma <jiam@marvell.com>
Tested-by: Jia Ma <jiam@marvell.com>
---
 include/smbios.h | 40 ++++++++++++++++++++++++++++++++++++++++
 lib/smbios.c     | 42 ++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 82 insertions(+)

diff --git a/include/smbios.h b/include/smbios.h
index e97cc3f977..031c9871e9 100644
--- a/include/smbios.h
+++ b/include/smbios.h
@@ -25,6 +25,7 @@ enum {
 
 #define NO_OF_CACHE CACHE_MAX
 #define MAX_PORTS 3
+#define MAX_SLOTS 2
 
 /* SMBIOS structure types */
 enum {
@@ -258,6 +259,45 @@ struct __packed smbios_type8 {
 	char eos[SMBIOS_STRUCT_EOS_BYTES];
 };
 
+/* Type 9 */
+#define DMTF_TYPE9_SLOT_TYPE_PCI			0x06
+#define DMTF_TYPE9_SLOT_TYPE_PCIE			0xA8
+#define DMTF_TYPE9_CURRENT_USAGE_AVAILABLE	0x03
+#define DMTF_TYPE9_SLOT_LENGTH_LONG			0x04
+#define DMTF_TYPE9_SLOT_CHAR_1_3_3V			BIT(2)
+#define DMTF_TYPE9_SLOT_DATA_BUS_WIDTH_X4	0x0A
+
+struct type9_data {
+	char slot_designation[16];
+	u8 slot_type;
+	u8 slot_data_bus_width;
+	u8 current_usage;
+	u8 slot_length;
+	u16 slot_id;
+	u8 slot_characteristics_1;
+};
+
+struct __packed smbios_type9 {
+	u8 type;
+	u8 length;
+	u16 handle;
+	u8 slot_designation;
+	u8 slot_type;
+	u8 slot_data_bus_width;
+	u8 current_usage;
+	u8 slot_length;
+	u16 slot_id;
+	u8 slot_characteristics_1;
+	u8 slot_characteristics_2;
+	u16 segment_group_number;
+	u8 bus_number;
+	u8 device_function_number_base;
+	u8 data_bus_width_base;
+	u8 peer_grouping_count;
+	u8 peer_groups;
+	char eos[SMBIOS_STRUCT_EOS_BYTES];
+};
+
 struct __packed smbios_type32 {
 	u8 type;
 	u8 length;
diff --git a/lib/smbios.c b/lib/smbios.c
index 50969a130c..822800fbdb 100644
--- a/lib/smbios.c
+++ b/lib/smbios.c
@@ -39,6 +39,11 @@ struct smbios_type8 type8_data[3] = {
 	},
 };
 
+struct type9_data plat_type9_data[MAX_SLOTS] = {
+	{"PCIe Slot 1", DMTF_TYPE9_SLOT_TYPE_PCIE, DMTF_TYPE9_SLOT_DATA_BUS_WIDTH_X4, DMTF_TYPE9_CURRENT_USAGE_AVAILABLE, DMTF_TYPE9_SLOT_LENGTH_LONG, 1, DMTF_TYPE9_SLOT_CHAR_1_3_3V},
+	{"PCIe Slot 2", DMTF_TYPE9_SLOT_TYPE_PCIE, DMTF_TYPE9_SLOT_DATA_BUS_WIDTH_X4, DMTF_TYPE9_CURRENT_USAGE_AVAILABLE, DMTF_TYPE9_SLOT_LENGTH_LONG, 2, DMTF_TYPE9_SLOT_CHAR_1_3_3V},
+};
+
 /**
  * smbios_add_string() - add a string to the string area
  *
@@ -370,6 +375,42 @@ static int smbios_write_type8(ulong *current, int handle)
 	return len;
 }
 
+static int smbios_write_type9_dm(ulong *current, int handle, int index)
+{
+	struct smbios_type9 *t;
+	int len = sizeof(struct smbios_type9);
+
+	t = map_sysmem(*current, len);
+	memset(t, 0, sizeof(struct smbios_type9));
+	fill_smbios_header(t, SMBIOS_SYSTEM_SLOTS, len, handle + index);
+
+	t->slot_designation = smbios_add_string(t->eos, plat_type9_data[index].slot_designation);
+
+	t->slot_id = plat_type9_data[index].slot_id;
+	t->slot_length = plat_type9_data[index].slot_length;
+	t->slot_type = plat_type9_data[index].slot_type;
+	t->slot_data_bus_width = plat_type9_data[index].slot_data_bus_width;
+
+	t->current_usage = plat_type9_data[index].current_usage;
+	t->slot_characteristics_1 = plat_type9_data[index].slot_characteristics_1;
+
+	len = t->length + smbios_string_table_len(t->eos);
+	*current += len;
+	unmap_sysmem(t);
+
+	return len;
+}
+
+static int smbios_write_type9(ulong *current, int handle)
+{
+	u32 no_of_handles = MAX_SLOTS, i = 0, len = 0;
+
+	for (; i < no_of_handles; i++)
+		len += smbios_write_type9_dm(current, handle, i);
+
+	return len;
+}
+
 static int smbios_write_type32(ulong *current, int handle)
 {
 	struct smbios_type32 *t;
@@ -408,6 +449,7 @@ static smbios_write_type smbios_write_funcs[] = {
 	smbios_write_type4,
 	smbios_write_type7,
 	smbios_write_type8,
+	smbios_write_type9,
 	smbios_write_type32,
 	smbios_write_type127
 };
-- 
2.29.0

